test/logging/timestamp_checkpoint_recovery_test.cpp

```diff
@@ -0,0 +1,600 @@
+          EXPECT_EQ(1, column.GetConstraints().size());
+          for (auto constraint : column.GetConstraints()) {
+            if (constraint.GetName() ==
+                "FK_checkpoint_constraint_test->checkpoint_index_test") {
+              EXPECT_EQ(ConstraintType::FOREIGN, constraint.GetType());
+              EXPECT_EQ(1, constraint.GetForeignKeyListOffset());
+              EXPECT_EQ(INVALID_OID, constraint.GetUniqueIndexOffset());
+            } else {
+              LOG_ERROR("Unexpected constraint is found: %s",
+                        constraint.GetName().c_str());
+            }
+          }
+        } else {
+          LOG_ERROR("Unexpected column is found: %s",
+                    column_catalog->GetColumnName().c_str());
+        }
+      } // loop end: single attribute constraint
+      // end: check the column constraint recovery
+    } else {
+      LOG_ERROR("Unexpected table is found: %s",
+                table_catalog->GetTableName().c_str());
+    }
+  } // table loop end
+
+  // check the catalog data
+  /*
+  auto catalog_db_catalog =
+      catalog->GetDatabaseObject(CATALOG_DATABASE_OID, txn);
+  for (auto table_catalog_pair : catalog_db_catalog->GetTableObjects()) {
+    auto table_catalog = table_catalog_pair.second;
+    // auto table = storage->GetTableWithOid(table_catalog->GetDatabaseOid(),
+    // table_catalog->GetTableOid());
+    LOG_DEBUG("Check catalog table %s", table_catalog->GetTableName().c_str());
+    // currently do nothing
+  }
+  */
+
+  // finish the low level test
+  txn_manager.CommitTransaction(txn);
+
+
+  // high level test
+  // make sure the records of 3 user tables created checkpointing test
+  // are correct and their constraints are working correctly through
+  // SQL execution.
+
+  // make sure the data in each user table is correct
+  std::string sql1 = "SELECT * FROM checkpoint_table_test;";
+  std::vector<std::string> expected1 = {"0|1.2|aaa", "1|12.34|bbbbbb",
+                                        "2|12345.7|ccccccccc", "3|0|xxxx"};
+  TestingSQLUtil::ExecuteSQLQueryAndCheckResult(sql1, expected1, false);
+
+  std::string sql2 = "SELECT * FROM checkpoint_index_test;";
+  std::vector<std::string> expected2 = {"1|2|3|4|5", "6|7|8|9|10",
+                                        "11|12|13|14|15"};
+  TestingSQLUtil::ExecuteSQLQueryAndCheckResult(sql2, expected2, false);
+
+  std::string sql3 = "SELECT * FROM checkpoint_constraint_test;";
+  std::vector<std::string> expected3 = {"1|2|3|4|0|1|2", "5|6|7|8|1|6|7",
+                                        "9|10|11|12|2|11|12"};
+  TestingSQLUtil::ExecuteSQLQueryAndCheckResult(sql3, expected3, false);
+
+  // make sure the constraints are working
+  // PRIMARY KEY (1 column: pid)
+  LOG_INFO("PRIMARY KEY (1 column) check");
+  std::string primary_key_dml1 =
+      "INSERT INTO checkpoint_table_test VALUES (0, 5.5, 'eee');";
+  ResultType primary_key_result1 =
+      TestingSQLUtil::ExecuteSQLQuery(primary_key_dml1);
```

Comment:
All the tests before this constraint `PRIMARY KEY (1 column) check` work well while a `Segmentation fault (core dumped)` is encountered here.

Reply:
This error occurs because `active_tile_groups` in DataTable object is not set correctly when tile groups are recovered.  The `active_tile_groups` is used by data insertion to find a tile group where the insertion can write the data. The number of tile groups in `active_tile_groups` depends on the number of CPU cores. (This is a reason why my local environment didn't make this error. If the number is just one, then the error didn't occur.) If there are multiple tile groups in `active_tile_groups`, a new tuple is inserted into a tile group decided by the number of existed tuples in the table like below:

```
  size_t active_tile_group_id = number_of_tuples_ % active_tilegroup_count_;
  tile_group = active_tile_groups_[active_tile_group_id];
  tuple_slot = tile_group->InsertTuple(tuple);
```

A new tile group is also set into `active_tile_groups` by the number of existed tuples. In recovery, when the second tile group is recovered, the recovered table has already had some tuples in the first recovered tile group. In this time, a place in `active_tile_groups` where the second tile group will be set might be in the same place where the first tile group is set. 

In Jenkins, the first tile group in `active_tile_groups` was overwritten by the second tile group. And the error occurred when the insertion tried to insert a tuple into another empty entry of `active_tile_groups` where the second tile group should exist originally.

I fixed the tile group insertion in our checkpoint recovery to avoid this overwriting.